### ISSUE TITLE
<Link /> components throws "Failed to execute 'removeChild' on 'Node'" when using next-intl with Nextjs' app router, parallel routes and route interception

### ISSUE THREAD

**Issue by DriesCruyskens** [amannn/next-intl#479]
### Description

The `<Link />` component throws an error when used with Nextjs' parallel routes and routes interception. A full refresh is necessary to clear the error.
The full error:

```
Unhandled Runtime Error
NotFoundError: Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node.
```
<details>
<summary>Call stack</summary>
removeChildFromContainer
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (34039:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20984:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21132:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21094:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21079:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21079:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21079:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21094:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21079:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21079:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21132:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21132:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21079:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21132:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21094:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21079:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21094:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21079:0)
recursivelyTraverseDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20904:0)
commitDeletionEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21079:0)
commitDeletionEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (20891:0)
recursivelyTraverseMutationEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21291:0)
commitMutationEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21366:0)
recursivelyTraverseMutationEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21305:0)
commitMutationEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21327:0)
recursivelyTraverseMutationEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21305:0)
commitMutationEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21758:0)
recursivelyTraverseMutationEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21305:0)
commitMutationEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21758:0)
recursivelyTraverseMutationEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21305:0)
commitMutationEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21758:0)
recursivelyTraverseMutationEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21305:0)
commitMutationEffectsOnFiber
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21758:0)
recursivelyTraverseMutationEffects
node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (21305:0)
</details>

### Mandatory reproduction URL (CodeSandbox or GitHub repository)

~~https://codesandbox.io/p/github/DriesCruyskens/next-intl-mre/main~~

https://github.com/DriesCruyskens/next-intl-mre

### Reproduction description

1. Open page (this opens the modal)
2. Changes locales
and/or:

1. Open page (this opens modal)
2. refresh page
3. Change locales

### Expected behaviour

I expect <Link /> component to direct me to the new locale instead of throwing an error.

**Comment by amannn** [comment]
I think your CodeSandbox is private, can you make it public?

<img width="753" alt="Screenshot 2023-08-28 at 09 53 45" src="https://github.com/amannn/next-intl/assets/4038316/049cc72c-ef43-48cb-8af4-4b16fd2a4d4b">

Is this a duplicate of https://github.com/amannn/next-intl/issues/255? So far my latest information is that Next.js currently has internal bugs that lead to issues when using parallel or intercepted routes with `next-intl`. Reproductions where `next-intl` is not used (but the behavior simulated) show the same errors, therefore this is unfortunately not actionable on the `next-intl` side and we have to wait for upstream fixes.

If your reproduction is able to demonstrate an error that is _caused_ by `next-intl` and is not related to an internal Next.js bug, I'd be happy to look into it.

**Comment by DriesCruyskens** [comment]
I think it is public but I can't access it either in private window.
<img width="297" alt="Screenshot 2023-08-28 at 11 10 28" src="https://github.com/amannn/next-intl/assets/26900172/c50ff13a-e4ac-44fc-a832-c5c24db7a566">

Here is the GitHub project instead: [https://github.com/DriesCruyskens/next-intl-mre](https://github.com/DriesCruyskens/next-intl-mre)

I never experienced this specific error in #255 so I thought I'd put it here in case others experience the same problem. I do understand that this seems like a waste of time since Nextjs is buggy still and the error message is cryptic. Parallel routes work fine without next-intl's <Link /> component though so I thought there must be something linked to next-intl specifically.

**Comment by amannn** [comment]
Thanks, that helps! 

When I switch the language on the modal page, I also see this warning on the console:

```
Warning: validateDOMNesting(...): <div> cannot appear as a child of <#document>.
```

My understanding is that the request for the modal route returns invalid markup that causes your app to break.

Can you try to remove all `next-intl` APIs from the components in a branch and see if you can still reproduce the error?

**Comment by DriesCruyskens** [comment]
> Can you try to remove all `next-intl` APIs from the components in a branch and see if you can still reproduce the error?

I can't reproduce since parallel routes + route interception works fine without i18n. However, if I use `next/link` or `useRouter()` (from next-intl) somehow the `const pathname = usePathname();` (from next-intl) fails to strip the language from the pathname causing 404 errors to `/fr-be/en-eu/events` for example.
Here's a reproduction: <https://github.com/DriesCruyskens/next-intl-mre/commit/7e1ef9e756f6ce1a5d7e3dce3bff5b0fc30b8ba9>

**Comment by DriesCruyskens** [comment]
I actually found the cause of the `validateDOMNesting` warning. It's thrown because the root layout does not contain html and body tags when trying to display a 404 page. After moving these tags to the root layout, navigating to a different locale (from a subroute) throws a 404 page. This works fine when navigating without locale change or hard navigating to subpages.
<https://github.com/DriesCruyskens/next-intl-mre/commit/43ac111311d86351e7f420821d033db95fa50fe3>

**Comment by amannn** [comment]
> I actually found the cause of the `validateDOMNesting` warning. It's thrown because the root layout does not contain html and body tags when trying to display a 404 page.

I found that you can render these from a root `not-found` page as well: https://github.com/amannn/next-intl/blob/main/examples/example-next-13/src/app/not-found.tsx

I also think that the bug is somehow related to the root layout not having an `html` tag, but we need to render this in `./[locale]/layout` to be able to add the `lang` attribute (see also [the reference implementation](https://nextjs.org/docs/app/building-your-application/routing/internationalization#static-generation) from Next.js).

Can the bug be summarized as parallel and intercepted routes not working properly when the root layout doesn't contain an `html` tag? In that case, I think it should be reported to Next.js, in reference to their i18n example in the docs.

**Comment by DriesCruyskens** [comment]
> Can the bug be summarized as parallel and intercepted routes not working properly when the root layout doesn't contain an html tag? In that case, I think it should be reported to Next.js, in reference to their i18n example in the docs.

Sadly I can't manage to create a minimal reproducible example with Nextjs alone (without next-intl). Parallel routes and intercepted routes on their own work fine when the root does not contain an html tag so I don't think they'll spent time on it either.

**Comment by amannn** [comment]
Hmm, I see. Can you point me to a commit where you have a working app with Next.js before and then in the commit it's visible which functionality from `next-intl` breaks the app? Ideally as minimal as possible, e.g. if we know that the app breaks once we add the middleware, but before any messages are used in the app, this would be really helpful.

**Comment by DriesCruyskens** [comment]
Commit where all works fine: <https://github.com/DriesCruyskens/next-intl-mre/commit/cc76436ada7f84c430eff7e73bef1beb45281e62>
The commit right after that redirects to 404 inside parallel route: <https://github.com/DriesCruyskens/next-intl-mre/commit/c0d10e97c3c9fb0cc14539f39827238f69a16b0c>

**Comment by amannn** [comment]
I'm sorry, but that doesn't seem to indicate what's going wrong here. You only encounter the issue when you click on a link, but in https://github.com/DriesCruyskens/next-intl-mre/commit/cc76436ada7f84c430eff7e73bef1beb45281e62 there's no link to be clicked on, so naturally nothing can break, right?

I tried replacing your links that are created with `next-intl/link` with this:

```tsx
import Link from 'next/link'

<Link className="underline" href="/en-eu/events/1">en-eu</Link>
<Link className="false" href="/fr-be/events/1">fr-be</Link>
<Link className="false" href="/nl-be/events/1">nl-be</Link>
```

There's no runtime error when clicking on a link, but the language isn't changed either.

That's what I tried to discuss above, that it would be necessary that we have a version of this that works without `next-intl`, so that we can isolate if the issue is related to Next.js or `next-intl`.
